pybotx_widgets/carousel.py

```diff
@@ -1,317 +1,324 @@
-
-    # Count new start position
-    content_len = len(widget_content)
-    if start_from < 0 or start_from > content_len:
-        start_from = abs(content_len - abs(start_from))
-
-    # Count end position
-    end = start_from + displayed_content_count
-
-    if len(widget_content) <= displayed_content_count:
-        # if all content displayed, then hide arrows
-        show_right_arrow = False
-        show_left_arrow = False
-    elif loop:
-        # Loop content and always show arrows
-        widget_content = cycle(widget_content)  # type: ignore
-        show_right_arrow = True
-        show_left_arrow = True
-    else:
-        show_right_arrow = end < content_len
-        show_left_arrow = start_from > 0
-
-    displayed_content = islice(widget_content, start_from, end)
-
-    if inline:
-        markup = _get_inline_markup(
-            message,
-            command,
-            displayed_content,
-            show_left_arrow,
-            show_right_arrow,
-            control_labels,
-            additional_markup,
-        )
-    else:
-        markup = _get_newline_markup(
-            message,
-            command,
-            displayed_content,
-            show_left_arrow,
-            show_right_arrow,
-            show_numbers,
-            PaginationInfo(start_from, end, displayed_content_count, content_len),
-            control_labels,
-            additional_markup,
-        )
-
-    await send_or_update_message(message, bot, label, markup=markup)
-
-    return None
 
+class ValidationMixin:
+    LEFT_ARROW: str
+    RIGHT_ARROW: str
 
-async def get_carousel_result(message: Message, bot: Bot) -> Optional[str]:
-    selected_val = message.data.get("carousel_selected_val", "")
-    selected_value_label = message.data.get("carousel_selected_value_label")
-    message_label = message.data.get("carousel_message_label")
+    widget_content: Sequence
+    SELECTED_VALUE_LABEL: str
+    _start_from: int
+    loop: bool
+    inline: bool
+    show_numbers: bool
 
-    # if user select value
-    if selected_val and selected_val not in {LEFT_PRESSED, RIGHT_PRESSED}:
-        await send_or_update_message(
-            message,
-            bot,
-            selected_value_label.format(label=message_label, selected_val=selected_val),
-        )
-        _clear_carousel_data(message)
+    def _is_valid_label_for_placing_numbers(self, label: str) -> bool:
+        number_places = list(string.Formatter().parse(label))
 
-        return selected_val
-
-    return None
-
-
-def _clear_carousel_data(message: Message) -> None:
-    """Clear widget data form message.data."""
-
-    message.data.pop("message_id", None)
-    message.data.pop("carousel_selected_val", None)
-    message.data.pop("carousel_selected_value_label", None)
-    message.data.pop("carousel_message_label", None)
-    message.data.pop("carousel_start_from", None)
+        if len([place for place in number_places if place[1] == ""]) != 2:
```

Comment:
Don't you think this expression is hard to read? Maybe move it above?

Reply:
remove that expression